mlflow/genai/optimize/adapt.py

```diff
@@ -0,0 +1,181 @@
+import uuid
+from concurrent.futures import ThreadPoolExecutor
+from typing import TYPE_CHECKING, Any, Callable
+
+import mlflow
+from mlflow.entities.model_registry import PromptVersion
+from mlflow.environment_variables import MLFLOW_GENAI_EVAL_MAX_WORKERS
+from mlflow.genai.evaluation.utils import (
+    _convert_eval_set_to_df,
+)
+from mlflow.genai.optimize.adapters import BasePromptAdapter, get_default_adapter
+from mlflow.genai.optimize.types import EvaluationResultRecord, LLMParams, PromptAdaptationResult
+from mlflow.genai.prompts import load_prompt, register_prompt
+from mlflow.genai.utils.trace_utils import convert_predict_fn
+from mlflow.utils import gorilla
+from mlflow.utils.annotations import experimental
+from mlflow.utils.autologging_utils.safety import _wrap_patch
+
+if TYPE_CHECKING:
+    from mlflow.genai.evaluation.utils import EvaluationDatasetTypes
+
+
+@experimental(version="3.5.0")
+def adapt_prompts(
+    predict_fn: Callable[..., Any],
+    train_data: "EvaluationDatasetTypes",
+    target_prompt_uris: list[str],
+    optimizer_lm_params: LLMParams,
+    optimizer: BasePromptAdapter | None = None,
+) -> PromptAdaptationResult:
+    """
+    This API optimizes prompts used in the passed in function to produce similar
+    outputs as the outputs in the dataset. This API can be used to maintain the
+    outputs of `predict_fn` when the language model used in the function is changed.
+
+    Args:
+        predict_fn: a target function to be optimized. The callable should receive inputs
+            as keyword arguments and return the response. The function should use
+            MLflow prompt registry and call `PromptVersion.format` during execution
```

Comment:
Do we also need the `predict_fn` to return the same type as the output data in `train_data`?

Reply:
Let me document that.